students/stefan_lund/Lesson_3/mailroom.py

```diff
@@ -0,0 +1,190 @@
+
+
+def accept_user_input(answer, accept):
+    """
+    "start" : "a", "b", "c"
+    "thank_you" : "list", "quit", anything else is interpreted to be a Name
+    "amount" : string of digits with or without decimals
+    """
+    if accept[0] == "start":
+        if len(answer) == 1 and answer.isalpha():
+            answer = answer.lower()
+            if answer in accept[1]:
+                return True
+        return False
+
+    if accept[0] == "thank_you":
+        if len(answer) == 4 and answer.isalpha():
+            answer = answer.lower()
+            if answer == accept[1] or answer == accept[2]:
+                return True, answer
+        return True, answer
+
+    if accept[0] == "amount":
+        if float(answer) <= 0:
+            return False
+        else:
+            if "." not in answer:
+                return answer.isdecimal()
+            elif "." in answer:
+                if answer.count('.') == 1:
+                    temp_answer = answer[:answer.index(".")] + answer[answer.index(".") + 1:]
+                    return temp_answer.isdecimal()
+
+
+def report(data):
+
+    name_len = 0
+    amount_len = 0
+    for i in range(0, len(data), 3):
+        name_i = i
+        amount_i = name_i + 1
+        if len(data[name_i]) > name_len:
+            name_len = len(data[name_i]) + 3
+        if len(str(data[amount_i])) > amount_len:
+            amount_len = len(str(data[amount_i])) + 3
+
+    temp_str = "Donor Name", "Total Given", "Num Gifts", "Average Gift"
+    temp_form = "\n{:<{nl}}|{:>{al}}|{:>{al}}|{:>{al}}"
+    header = temp_form.format(*temp_str, nl = name_len, al = amount_len)
+    print(header)
+    print("-" * (name_len + 3 * amount_len + 3))
+
+    for i in range(0, len(data), 3):
+        temp_form = "{:<{nl}} ${:>{al}.2f}  {:>{al}} ${:>{al}.2f}"
+        d0 = data[i]
+        d1 = data[i + 1]
+        d2 = data[i + 2]
+        if d2 == 0:
+            d3 = 0
+        else:
+            d3 = data[i + 1] / data[i + 2]
+        line = temp_form.format(d0, d1, d2, d3, nl = name_len, al = amount_len - 1)
+        print(line)
+
+
+def run_mailroom(data):
+    """
+    data: list with name, total of donations - dddd.cc, number of gift occations
+    """
+    # data = []
+    start_options = ["start", "Send a Thank You", "Create a Report", "quit"]
+    question = user_input
+    not_finished = True
+    while not_finished:
+        wrong_answer = True
+        while wrong_answer:
+            answer = question(start_options, data)
+            accept = ["start", ["a", "b", "c"]]
+            if accept_user_input(answer, accept):
+                answer = answer.lower()
+                wrong_answer = False
+
+        if answer == "a":
+            thank_you(data)
+        elif answer == "b":
+            report(data)
+        elif answer == "c":
+            not_finished = False
+        else:
+            print("We have a bug!")
+            not_finished = False
+    print("Leaving the Mailroom.")

```

Comment:
suggestion to simplify this entry point, the logic is a little over-complicated.

please review suggestions here:
https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom_tutorial.html#controlling-main-program-flow